### ISSUE TITLE
Catch self-referential links

### ISSUE THREAD

**Issue by eread** [rvben/rumdl#769]
### Problem or motivation

At the moment, if you assume a file called `test.md`, this Markdown is accepted by `rumdl`:

```markdown
# Markdown file heading

Some introductory text that has a [self-referential link](test.md) and another
[self-referential link to a heading](test.md#level-2-heading).

## Level 2 heading

Sub-section content.
```

However, you might want to:

- Avoid the first link altogether. The reader doesn't want to follow a link back to the file they're reading.
- Avoid a link fragment in the current file that refers also to the current file, because that might refresh the page before jumping to the section.

`rumdl` already has an "avoid unnecessary path traversal" rule in MD057 with the optional `compact-paths` setting: https://github.com/rvben/rumdl/blob/0610735e0d36bc1cfd5a77cc720c08132c241182/docs/md057.md#compact-paths. This feature request is similar to that, but that setting doesn't catch the instances referred to above.

### Proposed solution

One possibility is to extend existing rules. So:

- To catch the first link in the sample Markdown, consider it a violation of `compact-paths` of MD057, much like if `test.md` was in the `test` directory, this would be a violation:
  ```markdown
  [self-referential link](../test/test.md)
  ```
- To catch the second link in the sample Markdown, add a new option to MD051 (https://github.com/rvben/rumdl/blob/0610735e0d36bc1cfd5a77cc720c08132c241182/docs/md051.md#configuration), which would disallow links with link anchors to reference the file they're in.

### Alternatives considered

An alternative to extending existing rules is a new rule that's dedicated to "self-referential links". There are already a heap of link-related rules, so I'm not sure if more rules are desirable.

On the other hand, "self-referential links" is a new class of problem that might deserve its own rule. It would be simpler to have one new rule than the functionality split between two existing rules.

### Feature category

Improve existing rule

### Example markdown

Assuming the Markdown is in a file called `test.md`:

```markdown
# Markdown file heading

Some introductory text that has a [self-referential link](test.md) and another
[self-referential link to a heading](test.md#level-2-heading).

## Level 2 heading

Sub-section content.
```

**Comment by rvben** [comment]
Shipped in v0.2.48, as one MD057 option rather than a split across two rules:

```toml
[MD057]
self-referential-links = true
```

On your sample, saved as `test.md`:

```
test.md:3:59: [MD057] Relative link 'test.md' points to the file it is in
test.md:4:38: [MD057] Relative link 'test.md#level-2-heading' points to the file it is in and can be simplified to '#level-2-heading' [*]
```

The fragment form is autofixable, so `rumdl fmt` rewrites it to `[...](#level-2-heading)`. The whole-file link is reported but not fixed: dropping the link and keeping its text, or pointing it somewhere else, are both reasonable and the tool should not choose for you.

Both live in MD057 because both are judgments about a link's path, which is what MD057 is about. MD051 asks whether a fragment resolves, and in your second link it does; what is wrong there is the redundant file qualifier in front of it. Default off.

Thanks @eread, including for arguing both sides of the new-rule question. The deciding factor was that a reader configuring this thinks about links to files, which is one subject, not two.
